### PR TITLE
feat: Ancilla support in CircuitBuilder

### LINKED ISSUE #854: Add/discard ancillae in `CircuitBuilder`
So we can build circuits like
image

### PR DESCRIPTION
**Pull request by aborgna-q**
Adds `add_ancilla` and `discard_ancilla` methods to the circuit builder, to modify the list of tracked wires.
Also adds extra info in the error messages.

Closes #854.

### CODE REVIEW

**Comment by aborgna-q** [comment]
Alternatively, we could just have two functions
```haskell
track_wire : Wire -> usize
untrack_wire : usize -> Wire
```
and manually call them after/before allocating or discarding,
```rust
let ancilla: Wire = circ.append_with_outputs(q_alloc(), [])?[0];
let ancilla: usize = circ.track_wire(ancilla);

...

let ancilla: Wire = circ.untrack_wire(ancilla);
circ.append_and_consume(q_discard(), [ancilla]);
```

**Review by ss2165 — CHANGES_REQUESTED** [review]
I much prefer the track/untrack methods - keeps it generic without getting too qubit-y

Will do detailed review after that change if you agree!

**Comment by aborgna-q** [comment]
Done, I dropped the composable calls in favour of track/untrack.

File: quantinuum-hugr/src/builder/circuit.rs [hunk 1]
```diff
@@ -208,4 +294,54 @@ mod test {
...
+            FunctionType::new(type_row![QB], type_row![QB]).into(),
+            |mut f_build| {
+                let mut circ = f_build.as_circuit(f_build.input_wires());
+                assert_eq!(circ.n_wires(), 1);
+
+                let [q0] = circ.tracked_units_arr();
+                let [ancilla] = circ
+                    .append_with_outputs::<CircuitUnit>(q_alloc(), [])?
+                    .try_into()
+                    .expect("Expected a single ancilla wire");
```
Comment (ss2165) [inline quantinuum-hugr/src/builder/circuit.rs:310]: Maybe we should have a panicking `as_arr()` utility on `Outputs` too? Can be follow up

File: quantinuum-hugr/src/utils.rs [hunk 2]
```diff
@@ -85,6 +85,10 @@ pub(crate) mod test_quantum_extension {
             )
             .unwrap();
 
+        extension
+            .add_op(SmolStr::new_inline("CZ"), "CZ".into(), two_qb_func())
+            .unwrap();
+
```
Comment (ss2165) [inline quantinuum-hugr/src/utils.rs:91]: worth adding this rather than just using CX?

**Review by ss2165 — COMMENTED** [review]
I think worth adding a quick test of the two error modes `InvalidWireIndex` and `MismatchedLinearInputs`

File: quantinuum-hugr/src/utils.rs (hunk 2 above)
Comment (aborgna-q) [reply quantinuum-hugr/src/utils.rs:91]: I wanted to match the motivating example in the test, but we can't do it without also adding a classically controlled operation.

I'll drop this.

File: quantinuum-hugr/src/builder/circuit.rs (hunk 1 above)
Comment (aborgna-q) [reply quantinuum-hugr/src/builder/circuit.rs:308]: https://github.com/CQCL/hugr/pull/871

File: quantinuum-hugr/src/builder/circuit.rs [hunk 3]
```diff
@@ -344,4 +334,46 @@ mod test {
 
         assert_matches!(build_res, Ok(_));
     }
+
+    #[test]
+    fn circuit_builder_errors() {
+        let _build_res = build_main(
+            FunctionType::new(type_row![QB, QB], type_row![QB, QB]).into(),
```
Comment (ss2165) [inline quantinuum-hugr/src/builder/circuit.rs:341]: nit: can use `new_endo`

**Review by ss2165 — APPROVED** [review]

**Pull request merged**